### PR TITLE
[iOS] Fixes Image load event in new arch

### LINKED ISSUE #44896: Old architecture i new architecture IOS onLoad Image prop
### Description

Hello, I am the maintainer of `react-native-svg`.

We tried to include the same `onLoad` feature in our Image element. However, we noticed that the data obtained from the 'onLoad' function varies between old architecture and new architecture on iOS devices. This difference is specifically seen on iOS devices only.
In old architecture, the ‘onLoad’ data seems to show the actual image size, while in new architecture, it displays information about size of the component returned by yoga, which can differ from the real image size.

### Steps to reproduce

```
1. git clone git@github.com:bohdanprog/react-native-image-onLoad-prop.git
2. cd react-native-image-onLoad-prop
3. yarn
4. yarn start
5. yarn ios
```

### React Native Version

0.75.0-nightly-20240612-fd618819c

### Affected Platforms

Runtime - iOS

### Areas

Fabric - The New Renderer

### Output of `npx react-native info`

```text
System:
  OS: macOS 14.4.1
  CPU: (11) arm64 Apple M3 Pro
  Memory: 104.17 MB / 18.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.9.0
    path: ~/.nvm/versions/node/v20.9.0/bin/node
  Yarn:
    version: 3.6.4
    path: /opt/homebrew/bin/yarn
  npm:
    version: 10.1.0
    path: ~/.nvm/versions/node/v20.9.0/bin/npm
  Watchman:
    version: 2024.04.08.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /usr/local/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.5
      - iOS 17.5
      - macOS 14.5
      - tvOS 17.5
      - visionOS 1.2
      - watchOS 10.5
  Android SDK: Not Found
IDEs:
  Android Studio: 2023.3 AI-233.14808.21.2331.11709847
  Xcode:
    version: 15.4/15F31d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.10
    path: /usr/bin/javac
  Ruby:
    version: 2.6.10
    path: /usr/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.75.0-nightly-20240611-5b3a32142
    wanted: 0.75.0-nightly-20240611-5b3a32142
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: true
```

### Stacktrace or Logs

```text
It logs with returned value from onLoad:
1. New architecture:

Running "reactNativeImageOnLoadProp" with {"rootTag":1,"initialProps":{"concurrentRoot":true},"fabric":true}
 (NOBRIDGE) LOG  RNImage:ios load JPEG image from url {"source": {"height": 852, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 393}, "target": 2}
```
2. Old architecture:
```
Running "reactNativeImageOnLoadProp" with {"rootTag":1,"initialProps":{"concurrentRoot":false}}
 LOG  RNImage:ios load JPEG image from url {"source": {"height": 25.333333333333332, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 25.333333333333332}, "target": 3}
```

### Reproducer

https://github.com/bohdanprog/react-native-image-onLoad-prop

### Screenshots and Videos

_No response_

### PR DESCRIPTION
**Pull request by zhongwuzw**
## Summary:

Fixes #44896. cc @cipolleschi 

## Changelog:

[IOS] [FIXED] - Fixes Image load event in new arch

## Test Plan:

#44896

### CODE REVIEW

File: packages/react-native/React/Fabric/Mounting/ComponentViews/Image/RCTImageComponentView.mm [hunk 1]
```diff
@@ -136,7 +136,10 @@ - (void)didReceiveImage:(UIImage *)image metadata:(id)metadata fromObserver:(con
     return;
   }
 
-  static_cast<const ImageEventEmitter &>(*_eventEmitter).onLoad(_state->getData().getImageSource());
+  auto imageSource = _state->getData().getImageSource();
```
Comment (NickGerleman) [inline packages/react-native/React/Fabric/Mounting/ComponentViews/Image/RCTImageComponentView.mm:139]: For someone not super familiar with this code, does this mean that the `ImageSource` stored as state is incorrect? Wondering if this change should happen here when emitting, or if we are storing wrong information earlier.

Comment (zhongwuzw) [reply packages/react-native/React/Fabric/Mounting/ComponentViews/Image/RCTImageComponentView.mm:139]: @NickGerleman Hi, `ImageSource` is used for layout, it's correct for the state. we misuse the info when emit load events. 
 
Android part:
https://github.com/facebook/react-native/blob/8b53d41a88c3c7c83183baa46be307102aa90d54/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/image/ReactImageView.java#L200-L205

**Comment by bohdanprog** [comment]
Hello, I looked at the output of the `onLoad` function. Can you help me understand why the size differs for the same image between iOS and Android platforms?
Image link: `https://image-placeholder.com/images/actual-size/75x75.png`
IOS:
```ts
RNImage:ios load PNG image from url with strict size {"source": {"height": 25.333333333333332, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 25.333333333333332}, "target": 3}
```
ANDROID:
```ts
RNImage:android load PNG image from url with strict size {"source": {"height": 76, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 76}}
```

**Comment by zhongwuzw** [comment]
> Hello, I looked at the output of the `onLoad` function. Can you help me understand why the size differs for the same image between iOS and Android platforms? Image link: `https://image-placeholder.com/images/actual-size/75x75.png` IOS:
> 
> ```ts
> RNImage:ios load PNG image from url with strict size {"source": {"height": 25.333333333333332, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 25.333333333333332}, "target": 3}
> ```
> 
> ANDROID:
> 
> ```ts
> RNImage:android load PNG image from url with strict size {"source": {"height": 76, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 76}}
> ```

@bohdanprog I thinks it's because Android returns the pixel but iOS returns the points, so iOS's width * scale == Android's width

**Comment by bohdanprog** [comment]
> > Hello, I looked at the output of the `onLoad` function. Can you help me understand why the size differs for the same image between iOS and Android platforms? Image link: `https://image-placeholder.com/images/actual-size/75x75.png` IOS:
> > ```ts
> > RNImage:ios load PNG image from url with strict size {"source": {"height": 25.333333333333332, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 25.333333333333332}, "target": 3}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ANDROID:
> > ```ts
> > RNImage:android load PNG image from url with strict size {"source": {"height": 76, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 76}}
> > ```
> 
> @bohdanprog I thinks it's because Android returns the pixel but iOS returns the points, so `iOS's width * scale == Android's width`

@zhongwuzw I tried to show the Image component using the size from onLoad, but they look very different.

**Comment by zhongwuzw** [comment]
> > > Hello, I looked at the output of the `onLoad` function. Can you help me understand why the size differs for the same image between iOS and Android platforms? Image link: `https://image-placeholder.com/images/actual-size/75x75.png` IOS:
> > > ```ts
> > > RNImage:ios load PNG image from url with strict size {"source": {"height": 25.333333333333332, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 25.333333333333332}, "target": 3}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ANDROID:
> > > ```ts
> > > RNImage:android load PNG image from url with strict size {"source": {"height": 76, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 76}}
> > > ```
> > 
> > 
> > @bohdanprog I thinks it's because Android returns the pixel but iOS returns the points, so `iOS's width * scale == Android's width`
> 
> @zhongwuzw I tried to show the Image component using the size from onLoad, but they look very different.

@bohdanprog Hi, I recommend you create a new issue so that we can investigate your issue.

**Review by WoLewicki — COMMENTED** [review]
We added quite similar thing in `react-native-svg` based on this PR: https://github.com/software-mansion/react-native-svg/pull/2293, and it seems to work pretty well. The only issue was already mentioned by @bohdanprog but it can be targeted in another issue.

**Comment by bohdanprog** [comment]
> > > > Hello, I looked at the output of the `onLoad` function. Can you help me understand why the size differs for the same image between iOS and Android platforms? Image link: `https://image-placeholder.com/images/actual-size/75x75.png` IOS:
> > > > ```ts
> > > > RNImage:ios load PNG image from url with strict size {"source": {"height": 25.333333333333332, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 25.333333333333332}, "target": 3}
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > ANDROID:
> > > > ```ts
> > > > RNImage:android load PNG image from url with strict size {"source": {"height": 76, "uri": "https://image-placeholder.com/images/actual-size/75x75.png", "width": 76}}
> > > > ```
> > > 
> > > 
> > > @bohdanprog I thinks it's because Android returns the pixel but iOS returns the points, so `iOS's width * scale == Android's width`
> > 
> > 
> > @zhongwuzw I tried to show the Image component using the size from onLoad, but they look very different.
> 
> @bohdanprog Hi, I recommend you create a new issue so that we can investigate your issue.

Here is a new issue: https://github.com/facebook/react-native/issues/45188

**Review by cipolleschi — APPROVED** [review]
The code looks good to me.

**Pull request closed without merging**